src/runtime/hexagon_remote/qurt_thread_pool.cpp

```diff
@@ -0,0 +1,271 @@
+                // release the lock and go to sleep.
+                // ***************************
+                // *** Work queue unlocked ***
+                // ***************************
+                work_queue.unlock();
+                qurt_sem_down(&wait_for_work);
+                 // ***********************
+                // *** Lock work queue ***
+                // ***********************
+                work_queue.lock();
+                continue;
+            }
+        } else {
+            // We are here only if this thread owns a job that is not done.
+            // So do a part of it.
+            if (owned_job->claimed()) {
+                // the owner goes to sleep till the workers wake it up.
+                qurt_cond_wait(&owned_job->wakeup_owner, &work_queue.work_mutex);
+                // This thread should have the lock now after having been woken up.
+                break;
+            } else {
+                job = owned_job;
+            }
+        }
+        int myjob = job->next++;
+        job->active_workers++;
+        // If all tasks of the job have been claimed, then pop the job off the stack.
+        if (job->claimed()) {
+            work_queue.jobs = job->next_job;
+        }
+        // ***************************
+        // *** Work queue unlocked ***
+        // ***************************
+        work_queue.unlock();
+
+        if (!locked) {
+            int lock_status = qurt_hvx_lock(job->curr_hvx_mode);
+            if (lock_status != QURT_EOK) {
+                halide_print(job->user_context, "QuRT: qurt_hvx_lock failed (thread pool)\n");
+            }
+        }
+
+        int result = halide_do_task(job->user_context, job->f,
+                                    myjob, job->closure);
+
+        qurt_hvx_unlock();
+        locked = false;
+
+        // ***********************
+        // *** Lock work queue ***
+        // ***********************
+        work_queue.lock();
+        if (result) {
+            job->exit_status = result;
+            job->next = job->end;
+            work_queue.jobs = job->next_job;
+        }
+        job->active_workers--;
+        if (job->done()) {
+            if (!owned_job) {
+                qurt_cond_signal(&(job->wakeup_owner));
+            }
+        }
+    }
+    // ***************************
+    // *** Work queue unlocked ***
+    // ***************************
+    work_queue.unlock();
+}
+
+void create_threads(int num_threads) {
+    // Acquire a lock on the work queue.
+    qurt_thread_attr_t thread_attr;
+    for (int i = 0; i < num_threads; ++i) {
+        qurt_thread_attr_init(&thread_attr);
+        qurt_thread_attr_set_stack_addr(&thread_attr, stack[i]);
+        qurt_thread_attr_set_stack_size(&thread_attr, STACK_SIZE);
+        qurt_thread_attr_set_priority(&thread_attr, 100);
+        qurt_thread_create(&threads[i], &thread_attr, (void (*)(void *))worker_thread, NULL);
+    }
+}
+
+void qurt_thread_pool_init() {
```

Comment:
This is not threadsafe. What prevents two kernels from being invoked on different threads of the same HVX processor?